### PR TITLE
Configure the ability to add event waiting time in gateway

### PR DESCRIPTION
**Pull request by Sarangan0219**
### Purpose
Related to wso2/product-apim#8246

### Approach

- This PR adds a new config of `event_waiting_time`. Gateway will wait for the time specified in this config after receiving an event. This is to ensure that DB retrieval is not ahead of DB updates thereby avoiding inconsistencies.
- Resolves issues in `data_retrieval_mode` config.
- Resolves issues in `deployment_retry_duration` config.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java [hunk 1]
```diff
@@ -81,9 +84,18 @@ public void onMessage(Message message) {
...
                              */
-                            handleNotificationMessage((String) map.get(APIConstants.EVENT_TYPE),
-                                    (Long) map.get(APIConstants.EVENT_TIMESTAMP),
-                                    (String) map.get(APIConstants.EVENT_PAYLOAD));
+
+                            ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+                            final Runnable task = new Runnable() {
+                                @Override
+                                public void run() {
+                                    handleNotificationMessage((String) map.get(APIConstants.EVENT_TYPE),
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:92]: I think putting it this way would make each and every event handled in a different thread. And be started with a delay. Can we only start a task when an API is need to be deployed.

Comment (jaadds) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:92]: We can do the change like below:
1. Create a new ThreadScheduler at the JMSMessageListener with a fixed number of threads.
2. While submitting the Task to the above Scheduler override the run method either with deploy or undeploy methods.
3. Use no delay while submitting the task - we initially discussed about it, but in a later discussion changed the approach.
4. Now the thread will be different from JMS event receiving thread. So do the waiting in the same thread by using a thread sleep and continue with the pulling and deployment.

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:97]: https://github.com/wso2/carbon-apimgt/pull/9016/commits/67da6203b4479b5838882a8416f61db2eb3af7fe

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java [hunk 2]
```diff
@@ -55,6 +58,7 @@
     private InMemoryAPIDeployer inMemoryApiDeployer = new InMemoryAPIDeployer();
     GatewayArtifactSynchronizerProperties gatewayArtifactSynchronizerProperties = ServiceReferenceHolder
             .getInstance().getAPIManagerConfiguration().getGatewayArtifactSynchronizerProperties();
+    ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(10);
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:61]: Make this a constant.

Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java]: As a practice give a name to Threads in the thread pool. So use the override of newScheduledThreadPool which accepts a ThreadFactory, and use it to give a name.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/DBRetriever.java [hunk 3]
```diff
@@ -60,6 +64,11 @@ public void init() throws ArtifactSynchronizerException {
     public String retrieveArtifact(String APIId, String gatewayLabel, String gatewayInstruction)
             throws ArtifactSynchronizerException {
         CloseableHttpResponse httpResponse = null;
+        try {
+            Thread.sleep(gatewayArtifactSynchronizerProperties.getEventWaitingTime());
+        } catch (InterruptedException e) {
+            // Ignore
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/DBRetriever.java:70]: Log the exception here.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java [hunk 4]
```diff
@@ -111,12 +114,24 @@ private void handleNotificationMessage(String eventType, long timestamp, String
...
-                    inMemoryApiDeployer.unDeployAPI(gatewayEvent.getApiId(), gatewayLabel);
-                }
+                    task = new Runnable() {
+                        @Override
+                        public void run() {
+                            inMemoryApiDeployer.unDeployAPI(gatewayEvent.getApiId(), gatewayLabel);
+                        }
+                    };
                 }
+                scheduler.schedule(task, 1, TimeUnit.MILLISECONDS);
```
Comment (jaadds) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:145]: Add a meaningful debug log correlating EventId, APIName and other useful details in the event.

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:145]: https://github.com/wso2/carbon-apimgt/pull/9016/commits/bb95666b85e9ca2fe650f54b212bcdce5d06008f

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/DBRetriever.java (hunk 3 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/DBRetriever.java]: https://github.com/wso2/carbon-apimgt/pull/9016/commits/bb95666b85e9ca2fe650f54b212bcdce5d06008f

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java (hunk 2 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java]: https://github.com/wso2/carbon-apimgt/pull/9016/commits/5e8ba5419c5fc20d7981d5a593400c8ceecf9d2b

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java [hunk 5]
```diff
@@ -111,24 +114,39 @@ private void handleNotificationMessage(String eventType, long timestamp, String
...
-                            log.debug(gatewayEvent.getEventId() + " processed and deployed in gateway");
+                Runnable task = null;
+                if (APIConstants.EventType.DEPLOY_API_IN_GATEWAY.name().equals(eventType)) {
+                    task = new Runnable() {
+
+                        @Override public void run() {
+                            try {
+                                inMemoryApiDeployer.deployAPI(gatewayEvent.getApiId(), gatewayLabel);
+                            } catch (ArtifactSynchronizerException e) {
+                                log.error("Error in deploying artifacts");
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:129]: Don't we have to include context information in the error log? ie: API id

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/listeners/GatewayJMSMessageListener.java:129]: Fixed https://github.com/wso2/carbon-apimgt/pull/9542/commits/63c98e787dbe2323e2fc0775fc72e5c5d9c7f030

**Pull request merged**